### PR TITLE
tpcc: use stateless retries to avoid deadlocks

### PR DESCRIPTION
**Pull request by miraradeva**
The TPCC workload uses a stateful retry mechanism
(`cockroach_restart` savepoint) which results in retaining the transaction's locks across retries. This have been known to increase the likelihood of deadlocks (see #134870).

This commit introduces a stateless retry loop for the TPCC workload, similar to the one in YCSB. In conjunction with #134870, it should help the failures we see in #131585.

Informs: #131585

Release note: None

### CODE REVIEW

File: pkg/workload/tpcc/tpcc.go [hunk 1]
```diff
@@ -1096,7 +1098,21 @@ func (w *tpcc) executeTx(
...
+			if err != nil {
+				return onTxnStartDuration, err
+			}
+			err = txnFuncWithStartFuncs(tx)
+			if err == nil {
+				return onTxnStartDuration, tx.Commit(ctx)
+			}
+			var pgErr *pgconn.PgError
+			if errors.As(err, &pgErr) && pgcode.MakeCode(pgErr.Code) == pgcode.SerializationFailure {
+				continue
```
Comment (srosenberg) [inline pkg/workload/tpcc/tpcc.go:1112]: Previous mechanism had an upper bound on the number of retries. (I believe the default is `50`.) Shouldn't we enforce it here?

**Comment by srosenberg** [comment]
> The TPCC workload uses a stateful retry mechanism
(cockroach_restart savepoint) which results in retaining the transaction's locks across retries. This have been known to increase the likelihood of deadlocks (see https://github.com/cockroachdb/cockroach/pull/134870).

Do we know what caused this regression? Afaik, deadlocks started only recently, whereas the previous retry mechanism had been running for a (long) while.

File: pkg/workload/tpcc/tpcc.go (hunk 1 above)
Comment (miraradeva) [reply pkg/workload/tpcc/tpcc.go:1112]: Yes, good point. I'll add the maxRetries back in.

**Comment by miraradeva** [comment]
> Do we know what caused this regression? Afaik, deadlocks started only recently, whereas the previous retry mechanism had been running for a (long) while.

This test was introduced in Feb (https://github.com/cockroachdb/cockroach/pull/118852); Nathan stabilized it in Apr by ensuring serializable transactions acquire durable locks (https://github.com/cockroachdb/cockroach/pull/122242). It failed a couple of times in July (https://github.com/cockroachdb/cockroach/issues/127811) and Sep (https://github.com/cockroachdb/cockroach/issues/130309), but then it started failing more regularly, starting in Sep/Oct (https://github.com/cockroachdb/cockroach/issues/131585). I don't have a good explanation for this; I'll let @arulajmani hypothesize too.

If we want to understand this a bit better, we can let https://github.com/cockroachdb/cockroach/pull/134870 merge in first and observe how this test behaves, before changing the retry mechanism.

**Review by arulajmani — APPROVED** [review]
:lgtm: once the `maxRetries` is added back in. 

> If we want to understand this a bit better, we can let [ workload/tpcc: re-order NewOrder txn locking #134870](https://reviewable.io/reviews/cockroachdb/cockroach/134870) merge in first and observe how this test behaves, before changing the retry mechanism.

I don't think it's worth waiting to merge this -- from Nathan's comment [here](https://github.com/cockroachdb/cockroach/pull/134870#issuecomment-2469250752), we know that we need to disable stateless restarts to fully stabilize this test.

> Reviewed 2 of 2 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/135316#-:-OBzqhZE38kkeZtFD-G5:bmr6jhu)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @DarrylWong, @srosenberg, and @vidit-bhat)

File: pkg/workload/tpcc/tpcc.go [hunk 2]
```diff
@@ -1096,7 +1101,22 @@ func (w *tpcc) executeTx(
 	}
 
 	if w.txnRetries {
-		return onTxnStartDuration, crdbpgx.ExecuteTx(ctx, conn, txOpts, txnFuncWithStartFuncs)
+		for i := 0; i <= maxRetries; i++ {
```
Comment (srosenberg) [inline pkg/workload/tpcc/tpcc.go:1104]: Sorry, I should have thought of this earlier... what about exposing the number of attempted retries as a metric, to help troubleshooting/diagnosing? On the other hand, I'm not aware of a specific metric for savepoints, which would have conveyed the same information.

**Review by srosenberg — APPROVED** [review]

File: pkg/workload/tpcc/tpcc.go (hunk 2 above)
Comment (miraradeva) [reply pkg/workload/tpcc/tpcc.go:1105]: Not sure about metrics, but I added a log line similar to [this](https://github.com/cockroachdb/cockroach/blob/c83c57d354741ac36740894f7387c014eb6c09fe/pkg/kv/txn.go#L1133). We should be able to see in the logs if a transaction is stuck retrying.

**Pull request merged**